vulnerabilities/importers/github.py

```diff
@@ -0,0 +1,342 @@
+# second '%s' is interesting, it will have the value '' for the first request,
+# since we don't have any value for endCursor at the beginning
+# for all the subsequent requests it will have value 'after: "{endCursor}""
+query = '''
+        query MyQuery {
+        securityVulnerabilities(first: 100, ecosystem: %s, %s) {
+            edges {
+            node {
+                advisory {
+                identifiers {
+                    type
+                    value
+                }
+                summary
+                }
+                package {
+                name
+                }
+                vulnerableVersionRange
+            }
+            }
+            pageInfo {
+            hasNextPage
+            endCursor
+            }
+        }
+        }
+        '''
+
+
+class GitHubTokenError(Exception):
+    pass
+
+
+@dataclasses.dataclass
+class GitHubAPIDataSourceConfiguration(DataSourceConfiguration):
+    endpoint: str
+    ecosystems: list
+
+
+class GitHubAPIDataSource(DataSource):
+
+    CONFIG_CLASS = GitHubAPIDataSourceConfiguration
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        try:
+            self.gh_token = os.environ['GH_TOKEN']
+        except KeyError:
+            raise GitHubTokenError('Envirnomental variable GH_TOKEN is missing')
+
+    def __enter__(self):
+        self.advisories = self.fetch()
+
+    def updated_advisories(self) -> Set[Advisory]:
+        return self.batch_advisories(self.process_response())
+
+    def fetch(self) -> Mapping[str, List[Mapping]]:
+
+        headers = {'Authorization': 'token ' + self.gh_token}
+        api_data = {}
+        for ecosystem in self.config.ecosystems:
+
+            api_data[ecosystem] = []
+            end_cursor_exp = ''
+
+            while True:
+
+                query_json = {'query': query % (ecosystem, end_cursor_exp)}
+                resp = requests.post(
+                    self.config.endpoint, headers=headers, json=query_json
+                ).json()
+
+                if resp.get('message') == 'Bad credentials':
+                    raise GitHubTokenError('Invalid GitHub token')
+
+                end_cursor = resp['data']['securityVulnerabilities']['pageInfo'][
+                    'endCursor'
+                ]
+                end_cursor_exp = 'after: {}'.format('"{}"'.format(end_cursor))
+                api_data[ecosystem].append(resp)
+                print(resp)
+
+                if not resp['data']['securityVulnerabilities']['pageInfo'][
+                    'hasNextPage'
+                ]:
+                    break
+        return api_data
+
+    def set_version_api(self, ecosystem: str) -> None:
+
+        if ecosystem == 'MAVEN':
+            self.version_api = MavenVersionAPI()
+
+        elif ecosystem == 'NUGET':
+            self.version_api = NugetVersionAPI()
+
+        elif ecosystem == 'COMPOSER':
+            self.version_api = ComposerVersionAPI()
+
```

Comment:
What about this instead?
```
    def set_version_api(self, ecosystem: str) -> None:
        versioners = dict(
            MAVEN=MavenVersionAPI,
            NUGET=NugetVersionAPI,
            COMPOSER=ComposerVersionAPI,
        )
        versioner = self.get_versioner(ecosystem)
        if versioner:
            self.version_api = versioner()
```
Using a mapping for long conditionals is a useful idiom :)  .... and here I expect that list to be growing, right?

Reply:
No, actually check https://github.com/nexB/vulnerablecode/issues/74#issuecomment-642041958

Reply:
Your approach is better, I will use that instead